src/GameLogic/DataStructures/ActionSequence.ts

```diff
@@ -4,21 +4,20 @@ export class ActionSequence {
     private _actionSequence: PlayerAction[];
     private _step: number = 0;
 
-	constructor(actionSequence: PlayerAction[] = [])
-	{
-		this._actionSequence = actionSequence;
+	constructor(actionSequence: PlayerAction[] = []) {
+		this._actionSequence = Array.from(actionSequence);
     }
     
     public push(action: PlayerAction): void {
 		this._actionSequence.push(action);
     }
     
-    public getNext(): PlayerAction {
+    public getNext(): PlayerAction | undefined {
         if (this._step >= this._actionSequence.length) {
             return undefined;
         }
-        let output = this._actionSequence[this._step];
-        this._step++;
-        return output;
+        return this._actionSequence[this._step++];
     }
+
+    public copyActions = () : PlayerAction [] => Array.from(this._actionSequence);
```

Comment:
I'd replace it with:
```
public get actions(): readonly PlayerAction[] {
    return this._actionSequence;
}
```

Since the method returns something it's better to have it in the name, copy suggest like it does some kind of an external operation. You also don't need to make a copy at this point, a readonly array will work as well (of someone wants to modify it in any way then they can make a copy at that point), this way you reduce the footprint of this method.